homeassistant/components/risco/__init__.py

```diff
@@ -18,17 +29,98 @@
 from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
 
-from .const import DATA_COORDINATOR, DEFAULT_SCAN_INTERVAL, DOMAIN, EVENTS_COORDINATOR
+from .const import (
+    DATA_COORDINATOR,
+    DEFAULT_SCAN_INTERVAL,
+    DOMAIN,
+    EVENTS_COORDINATOR,
+    PARTITION_UPDATES,
+    SYSTEM,
+    TYPE_LOCAL,
+    ZONE_UPDATES,
+)
 
 PLATFORMS = [Platform.ALARM_CONTROL_PANEL, Platform.BINARY_SENSOR, Platform.SENSOR]
-UNDO_UPDATE_LISTENER = "undo_update_listener"
+UNDO_LISTENERS = "undo_listeners"
 LAST_EVENT_STORAGE_VERSION = 1
 LAST_EVENT_TIMESTAMP_KEY = "last_event_timestamp"
 _LOGGER = logging.getLogger(__name__)
 
 
+def is_local(entry: ConfigEntry) -> bool:
+    """Return whether the entry represents an instance with local communication."""
+    return entry.data.get(CONF_TYPE) == TYPE_LOCAL
+
+
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up Risco from a config entry."""
+    if is_local(entry):
+        return await _async_setup_local_entry(hass, entry)
+
+    return await _async_setup_cloud_entry(hass, entry)
+
+
+async def _async_setup_local_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    data = entry.data
+    risco = RiscoLocal(data[CONF_HOST], data[CONF_PORT], data[CONF_PIN])
+
+    try:
+        await risco.connect()
+    except CannotConnectError as error:
+        raise ConfigEntryNotReady() from error
+    except UnauthorizedError:
+        _LOGGER.exception("Failed to login to Risco cloud")
+        return False
+
+    async def _error(error):
+        _LOGGER.error("Error in Risco library: %s", error)
+
+    remove_error = risco.add_error_handler(_error)
+
+    async def _default(command, result, *params):
+        _LOGGER.debug(
+            "Unhandled update from Risco library: %s, %s, %s", command, result, params
+        )
+
+    remove_default = risco.add_default_handler(_default)
+
+    zone_updates: dict[int, Callable[[], Any]] = {}
+
+    async def _zone(zone_id, zone):
+        _LOGGER.debug("Risco zone update for %d", zone_id)
+        callback = zone_updates.get(zone_id)
+        if callback:
+            callback()
+
+    remove_zone = risco.add_zone_handler(_zone)
+
+    partition_updates: dict[int, Callable[[], Any]] = {}
+
+    async def _partition(partition_id, partition):
+        _LOGGER.debug("Risco partition update for %d", partition_id)
+        callback = partition_updates.get(partition_id)
+        if callback:
+            callback()
+
+    remove_partition = risco.add_partition_handler(_partition)
+
+    listenrs = [remove_error, remove_default, remove_zone, remove_partition]
+
+    listenrs.append(entry.add_update_listener(_update_listener))
+
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = {
+        UNDO_LISTENERS: listenrs,
+        ZONE_UPDATES: zone_updates,
+        PARTITION_UPDATES: partition_updates,
+        SYSTEM: risco,
+    }
```

Comment:
I recommend using a dataclass to store multiple instances in hass.data under a single key. That gives type annotations for free.

Reply:
What do you mean for free? Wouldn't the type definitions simply move to the class?

Reply:
The attributes will be typed once we know that we've got this dataclass from hass.data. Instead of needing to type all the items all the time, after looking up hass.data, we just need to type the dataclass after looking that up in hass.data.